backends/p4tools/testgen/targets/bmv2/target.cpp

```diff
@@ -0,0 +1,117 @@
+#include "backends/p4tools/testgen/targets/bmv2/target.h"
+
+#include <stddef.h>
+
+#include <map>
+#include <string>
+#include <vector>
+
+#include "ir/ir.h"
+#include "lib/cstring.h"
+#include "lib/exceptions.h"
+#include "lib/ordered_map.h"
+
+#include "backends/p4tools/testgen/lib/namespace_context.h"
+#include "backends/p4tools/testgen/targets/bmv2/constants.h"
+#include "backends/p4tools/testgen/targets/bmv2/test_backend.h"
+
+namespace P4Tools {
+
+namespace P4Testgen {
+
+namespace Bmv2 {
+
+/* =============================================================================================
+ *  BMv2_V1ModelTestgenTarget implementation
+ * ============================================================================================= */
+
+BMv2_V1ModelTestgenTarget::BMv2_V1ModelTestgenTarget() : TestgenTarget("bmv2", "v1model") {}
+
+void BMv2_V1ModelTestgenTarget::make() {
+    static BMv2_V1ModelTestgenTarget* INSTANCE = nullptr;
+    if (INSTANCE == nullptr) {
+        INSTANCE = new BMv2_V1ModelTestgenTarget();
+    }
+}
+
+const BMv2_V1ModelProgramInfo* BMv2_V1ModelTestgenTarget::initProgram_impl(
+    const IR::P4Program* program, const IR::Declaration_Instance* mainDecl) const {
+    // The blocks in the main declaration are just the arguments in the constructor call.
```

Comment:
I actually would very much like to automate this step -- see https://github.com/p4lang/p4c/pull/3281
This should make it much easier to write tools specific to a model.

Reply:
Maybe if you have no other project going on we could team up?

Reply:
Yes, something like this is sorely needed for testgen because we have to map the structure of the package to the nodes of the parsed P4 program.. But I am not entirely sure how we would consume this model yet. 

I am happy to contribute to this, but I unfortunately have little time.